### ISSUE TITLE
About SFT baseline

### ISSUE THREAD

**Issue by hyhy892** [siyan-zhao/OPSD#7]
Thank you for your excellent work! When I attempted to reproduce the SFT baseline method described in the paper, I observed a significant drop in performance compared to Qwen3-4B. Specifically, at step 1000, the AIME24 avg@16 score was 59.38%; at step 1800, it was 58.13%; whereas the Qwen3-4B model achieved an avg@16 of 71.88%.
I used the `run_sft.sh` script provided in the code repository and performed training on eight A800 GPUs.
```bash
accelerate launch \
    --config_file accelerate.yaml \
    --num_processes "$NUM_PROCESSES" \
    --gradient_accumulation_steps 4 \
    --main_process_port 19346 \
    sft_train.py \
    --model_name_or_path /xxx/Models/Qwen3-4B \
    --learning_rate 5e-6 \
    --per_device_train_batch_size 2 \
    --gradient_accumulation_steps 4 \
    --output_dir /xxx/qwen3-4b-4epochs-30k \
    --num_train_epochs 4 \
    --gradient_checkpointing \
    --use_peft \
    --lora_r 64 \
    --lora_alpha 128 \
    --lora_target_modules q_proj k_proj v_proj o_proj gate_proj up_proj down_proj \
    --max_length 16000 \
    --logging_steps 5 \
    --save_steps 100

```
My review parameters are as follows:

<img width="862" height="276" alt="Image" src="https://github.com/user-attachments/assets/bbd5d1de-c078-4a9d-8610-0cf6ef16f7fa" />
Below are my evaluation test cases; the model generated outputs in a non-thinking mode.

`
 "full_generation": "<think>\n\n</think>\n\n<|begin_of_thought|>\n\nOkay, let's try to tackle this problem step by step. So, Aya walks 9 kilometers every morning, stops at a coffee shop, and the total time she takes depends on her walking speed. When she walks at a constant speed of s km/h, the entire journey (including coffee shop) takes 4 hours. When she walks faster, at s + 2 km/h, the total time is 2 hours and 24 minutes. We need to find out how long the walk takes her if she walks at s + 1/2 km/h, including the coffee shop time t minutes.\n\nFirst, let's parse the given information. The key points are:\n\n1. Aya walks 9 km each morning.\n2. Her speed is either s km/h or s + 2 km/h.\n3. The total time includes both walking and time in the coffee shop (which is a fixed t minutes).\n4. When her speed is s km/h, total time is 4 hours.\n5. When her speed is s + 2 km/h, total time is 2 hours and 24 minutes (which is 2.4 hours).\n6. We need to find the total time when her speed is s + 0.5 km/h.\n\nSo, the problem is essentially about setting up equations based on the relationship between speed, time, and distance. Let's denote:\n\n- t as the time spent at the coffee shop in minutes.\n- For the first scenario (speed = s km/h), total time is 4 hours.\n- For the second scenario (speed = s + 2 km/h), total time is 2 hours and 24 minutes (which is 2.4 hours).\n\nSince the coffee shop time is the same in both cases (t minutes), we can set up equations where the walking time plus t minutes equals the total time. Let's convert the total times into hours for consistency with speed units (km/h).\n\nFirst scenario:\nTotal time = 4 hours = walking time + t/60 hours.\n\nSecond scenario:\nTotal time = 2.4 hours = walking time + t/60 hours.\n\nLet's define walking time as the time spent walking the 9 km. For the first scenario, walking time = distance / speed = 9 / s hours. So, the equation becomes:\n\n9/s + t/60 = 4.  [Equation 1]\n\nFor the second scenario, walking speed is s + 2 km/h, so walking time = 9 / (s + 2) hours. Then:\n\n9/(s + 2) + t/60 = 2.4.  [Equation 2]\n\nNow we have two equations with two variables: s and t. Let's try to solve for s and t.\n\nFirst, let's write both equations:\n\n1. 9/s + t/60 = 4\n2. 9/(s + 2) + t/60 = 2.4\n\nWe can subtract the second equation from the first to eliminate t/60:\n\n(9/s - 9/(s + 2)) + (t/60 - t/60) = 4 - 2.4\n\nSo, simplifying:\n\n9(1/s - 1/(s + 2)) = 1.6\n\nCompute the left side:\n\n1/s - 1/(s + 2) = [ (s + 2) - s ] / [ s(s + 2) ] = 2 / [s(s + 2) ]\n\nTherefore:\n\n9 * [2 / (s(s + 2))] = 1.6\n\nSimplify:\n\n18 / [s(s + 2)] = 1.6\n\nMultiply both sides by s(s + 2):\n\n18 = 1.6 * s(s + 2)\n\nDivide both sides by 1.6:\n\n18 / 1.6 = s(s + 2)\n\nCalculate 18 divided by 1.6. Let's do that:\n\n1.6 * 11 = 17.6, so 18 / 1.6 = 11.25 (since 18 - 17.6 = 0.4, 0.4 / 1.6 = 0.25; 11.25). Alternatively, 18 / 1.6 = (180 / 10) / (16 / 10) = 180 / 16 = 11.25.\n\nSo, 11.25 = s(s + 2)\n\nThus, s² + 2s - 11.25 = 0\n\nThis is a quadratic equation. Let's solve for s.\n\nUsing the quadratic formula:\n\ns = [ -2 ± sqrt( (2)^2 + 4*1*11.25 ) ] / (2*1)\n\nCompute discriminant:\n\nsqrt(4 + 45) = sqrt(49) = 7\n\nSo,\n\ns = [ -2 ± 7 ] / 2\n\nWe discard the negative solution because speed can't be negative.\n\ns = ( -2 + 7 ) / 2 = 5 / 2 = 2.5\n\nSo, s = 2.5 km/h.\n\nNow, substitute s back into one of the equations to find t. Let's use Equation 1:\n\n9/s + t/60 = 4\n\nPlug in s = 2.5:\n\n9 / 2.5 + t/60 = 4\n\nCompute 9 / 2.5: 9 / 2.5 = 3.6\n\nSo,\n\n3.6 + t/60 = 4\n\nSubtract 3.6:\n\nt/60 = 0.4\n\nMultiply by 60:\n\nt = 24 minutes.\n\nSo, the time spent in the coffee shop is 24 minutes. Let's verify this with Equation 2 to ensure consistency.\n\nEquation 2: 9/(s + 2) + t/60 = 2.4\n\ns + 2 = 2.5 + 2 = 4.5 km/h\n\nCompute 9 / 4.5 = 2\n\nSo,\n\n2 + 24/60 = 2 + 0.4 = 2.4, which matches the given total time.\n\nGreat, so our values for s and t are consistent.\n\nNow, the problem asks for the time when Aya walks at s + 0.5 km/h. So, s + 0.5 = 2.5 + 0.5 = 3 km/h.\n\nSo, her new speed is 3 km/h. Let's compute her walking time and total time.\n\nWalking time = 9 km / 3 km/h = 3 hours.\n\nTotal time = walking time + t/60 = 3 hours + 24 minutes = 3 hours + 0.4 hours = 3.4 hours.\n\nConvert 3.4 hours to minutes: 3 hours = 180 minutes, 0.4 hours = 24 minutes. So total time is 180 + 24 = 204 minutes.\n\nWait, but 3.4 hours is indeed 204 minutes. Let me check:\n\n3 hours is 180 minutes. 0.4 hours is 0.4 * 60 = 24 minutes. 180 + 24 = 204. Correct.\n\nBut wait, let me rephrase the problem statement to make sure I didn't misinterpret anything.\n\nThe problem says when she walks at s + 2 km/h, the total time is 2 hours and 24 minutes. Then we found s = 2.5, so s + 2 is 4.5 km/h, and total time is 2.4 hours (144 minutes). That checks out.\n\nNow, if she walks at s + 0.5 = 3 km/h, then walking time is 9 / 3 = 3 hours. Coffee shop time is still 24 minutes. So total time is 3 hours and 24 minutes. Converting that to minutes: 3*60 +24=180+24=204 minutes. So 204 minutes is the answer.\n\nBut let me just verify all steps once again to ensure there's no mistake.\n\nFirst, solving for s and t:\n\nEquation 1: 9/s + t/60 = 4\n\nEquation 2: 9/(s + 2) + t/60 = 2.4\n\nSubtracting equation 2 from equation 1 gives:\n\n9/s - 9/(s + 2) = 4 - 2.4 = 1.6\n\nThen:\n\n9*(1/s - 1/(s + 2)) = 1.6\n\nCompute 1/s -1/(s + 2) = 2/(s(s + 2))\n\nThus:\n\n9*2/(s(s + 2)) = 1.6\n\n18/(s(s + 2))=1.6\n\ns(s + 2)=18/1.6=11.25\n\nThen s² + 2s -11.25=0\n\nSolution: s=( -2 +7)/2=5/2=2.5, correct.\n\nThen t=24, correct.\n\nThen when speed is 3 km/h: 9/3=3h, total time 3h +24min=204min. Correct.\n\nBut let me check if there's another way to approach this problem, just to confirm.\n\nAlternatively, maybe express t from one equation and substitute into the other.\n\nFrom Equation 1:\n\nt = 60*(4 - 9/s)\n\nFrom Equation 2:\n\nt = 60*(2.4 - 9/(s +2))\n\nSet them equal:\n\n60*(4 - 9/s) = 60*(2.4 - 9/(s +2))\n\nCancel 60:\n\n4 - 9/s = 2.4 - 9/(s +2)\n\nThen:\n\n4 - 2.4 = 9/s - 9/(s +2)\n\n1.6 = 9*(1/s -1/(s +2)) same as before. So same steps, same result. So t is 24.\n\nThus, the total time when walking at 3 km/h is 3h +24min=204min.\n\nBut the problem asks to \"Find the number of minutes the walk takes her, including the t minutes spent in the coffee shop.\"\n\nSo, 204 minutes. Therefore, the answer is 204. But let's check once more to ensure.\n\nWait, another way: Let's compute the walking time at 3 km/h, which is 9/3=3 hours. Then add t=24 minutes, which is 0.4 hours. So total time is 3.4 hours. 3.4*60=204 minutes. Correct.\n\nAlternatively, compute total time as 3 hours and 24 minutes. 3*60=180, 180+24=204. Yep.\n\nSo, all checks confirm that the answer is 204 minutes.\n\n**Final Answer**\n\\boxed{204}\n\n<|end_of_thought|>\n\n<|begin_of_solution|>\n\nAya walks a 9-kilometer journey every morning, including time spent at a coffee shop. We need to determine the total time she spends, including the coffee shop, if she walks at \\( s + \\frac{1}{2} \\) km/h. \n\nGiven:\n1. When walking at \\( s \\) km/h, the total time is 4 hours.\n2. When walking at \\( s + 2 \\) km/h, the total time is 2 hours and 24 minutes (2.4 hours).\n\nLet \\( t \\) be the time spent at the coffee shop in minutes. \n\nWe set up the equations for the total time in hours:\n- For \\( s \\) km/h: \\( \\frac{9}{s} + \\frac{t}{60} = 4 \\)\n- For \\( s + 2 \\) km/h: \\( \\frac{9}{s + 2} + \\frac{t}{60} = 2.4 \\)\n\nSubtracting the second equation from the first to eliminate \\( \\frac{t}{60} \\):\n\\[\n\\left( \\frac{9}{s} - \\frac{9}{s + 2} \\right) = 4 - 2.4\n\\]\n\\[\n\\frac{18}{s(s + 2)} = 1.6\n\\]\n\\[\n18 = 1.6 \\cdot s(s + 2)\n\\]\n\\[\ns(s + 2) = 11.25\n\\]\n\nSolving the quadratic equation \\( s^2 + 2s - 11.25 = 0 \\):\n\\[\ns = \\frac{-2 + \\sqrt{4 + 45}}{2} = \\frac{5}{2} = 2.5\n\\]\n\nSubstituting \\( s = 2.5 \\) back into the first equation to find \\( t \\):\n\\[\n\\frac{9}{2.5} + \\frac{t}{60} = 4\n\\]\n\\[\n3.6 + \\frac{t}{60} = 4\n\\]\n\\[\n\\frac{t}{60} = 0.4 \\implies t = 24\n\\]\n\nNow, if Aya walks at \\( s + \\frac{1}{2} = 3 \\) km/h, the walking time is:\n\\[\n\\frac{9}{3} = 3 \\text{ hours}\n\\]\n\nTotal time including the coffee shop:\n\\[\n3 \\text{ hours} + 24 \\text{ minutes} = 3 \\text{ hours} + 0.4 \\text{ hours} = 3.4 \\text{ hours}\n\\]\nConverting to minutes:\n\\[\n3.4 \\times 60 = 204 \\text{ minutes}\n\\]\n\nThus, the number of minutes the walk takes her, including the coffee shop time, is \\(\\boxed{204}\\).\n\n<|end_of_solution|>",
`

**Comment by siyan-zhao** [comment]
Hi, thanks for your report! We also see SFT baseline decrease performance significantly, as the solution it is trained on is very concise for a thinking model (see the "solution" column at https://huggingface.co/datasets/siyanzhao/Openthoughts_math_30k_opsd/viewer/default/train), and so we SFT it in non-thinking mode as it is not lengthy thinking reasoning trace. This makes the reasoning concise and decreases the score. 

Please see the updated paper https://arxiv.org/pdf/2601.18734v3 's Table 7, we only report the best score for SFT within 100 steps of training (e.g. 50, 75 steps), the same number of examples seen as OPSD.

**Comment by walawalagoose** [comment]
> Hi, thanks for your report! We also see SFT baseline decrease performance significantly, as the solution it is trained on is very concise for a thinking model (see the "solution" column at https://huggingface.co/datasets/siyanzhao/Openthoughts_math_30k_opsd/viewer/default/train), and so we SFT it in non-thinking mode as it is not lengthy thinking reasoning trace. This makes the reasoning concise and decreases the score.
> 
> Please see the updated paper https://arxiv.org/pdf/2601.18734v3 's Table 7, we only report the best score for SFT within 100 steps of training (e.g. 50, 75 steps), the same number of examples seen as OPSD.

Sorry for interuption. Does this mean that we should set `enable_thinking=false` during the SFT training of Qwen3-4B?

By the way, we find that with checkpoints at step=100, only the 4B model will encounter this performance drop; 1B and 8B won't. May I ask did you encounter the same phenomenon?

Thanks for the great work!

**Comment by siyan-zhao** [comment]
Hi! @walawalagoose  the current SFT text is already in non-thinking mode, you can check the printing to see. In my case the 1B model's performance decreased at 100 steps. Just one quick check, are you using zero2 or zero3?

**Comment by walawalagoose** [comment]
> Hi! [@walawalagoose](https://github.com/walawalagoose) the current SFT text is already in non-thinking mode, you can check the printing to see. In my case the 1B model's performance decreased at 100 steps. Just one quick check, are you using zero2 or zero3?

Thanks for the rapid reply! About `enable_thinking` in SFT, shall we set it to false in evaluation as well?

As for the performance drop, here are the experimental results we reproduced below:

| Method         | AIME24 | AIME25 | HMMT25 | Average |
| -------------- | ------ | ------ | ------ | ------- |
| **Qwen3-1.7B** |        |        |        |         |
| Base           | 51.1   | 36.9   | 24.2   | 37.4    |
| SFT (step=100) | 50.6   | 36.7   | 23.3   | 36.9    |
| **Qwen3-4B**   |        |        |        |         |
| Base           | 73.9   | 69.7   | 43.3   | 62.3    |
| SFT (step=100) | 21.9   | 22.5   | 11.7   | 18.7    |
| **Qwen3-8B**   |        |        |        |         |
| Base           | 75.3   | 66.1   | 45.0   | 62.1    |
| SFT (step=100) | 75.3   | 65.6   | 45.6   | 62.1    |

As shown, 1B and 8B models only encountered slight performance drop, while 4B model almost collapsed. This is quite an unusual phenomenon. 

We use the default zero2 config provided in `accelerate.yaml`.

**Comment by siyan-zhao** [comment]
Hi @walawalagoose, we didn't encounter the sharp drop for 4b, but I do see similar accuracy drop as @hyhy892 reported above with ~5-10% decrease on aime24 qwen3-4b. It could be a gradient spike during training or the lora adapter weights not being properly loaded? For evaluation, we enable thinking in eval for all models.

**Comment by sdws259** [comment]
I suspect it might have something to do with dataset shuffling? I noticed that the paper only evaluates the model at step 100, at which point the model hasn't finished a full epoch yet. It seems the code doesn't fix the sampling of training data – i.e., there's no fixed seed or deterministic data order – so the baseline would be trained on different subsets of the data across runs. That looks quite strange to me.

**Comment by JohnneyQin** [comment]
Thank you for your excellent paper!
I encounter this same issue while attempting to reproduce the SFT results for the 1.7B model: when evaluating the checkpoint at the end of 1 epoch, the 1.7B model's avg@12 score on AIME24 dropped from 48.33% to 29.72%. As the authors noted, the reported figures represent the best SFT results from a very early stage (the first 100 steps). Could you please provide performance data for multiple checkpoints, such as at the end of 1 epoch, as well as at 100, 200, and 500 steps?
I look forward to your reply!
